src/control/lib/netdetect/netdetect.go

```diff
@@ -0,0 +1,163 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+package netdetect
+/*
+#cgo CFLAGS: -I${SRCDIR}/../../include
+#cgo LDFLAGS: -lhwloc
+#include <stdlib.h>
+#include <hwloc.h>
+#include <stdio.h>
+*/
+import "C"
+import "unsafe"
+
+import (
+	"github.com/daos-stack/daos/src/control/log"
+	"github.com/pkg/errors"
+	"net"
+	"strings"
+)
+
+// initLib initializes the hwloc library.  For testing purposes,
+// set xmlTopology to a valid XML topology file name (generated by lstopo)
+// and hwloc will make queries against the data in this file.  Otherwise,
+// specify no topology file and hwloc will perform queries against the actual
+// local system.
+func initLib(xmlTopology string) (C.hwloc_topology_t, error) {
+	var topology C.hwloc_topology_t
+
+	status := C.hwloc_topology_init(&topology)
+	if status != 0 {
+		return nil, errors.New("hwloc_topology_init failure")
+	}
+
+	if (xmlTopology != "") {
+		status := C.hwloc_topology_set_xml(topology, C.CString(xmlTopology))
+		if (status != 0) {
+			return nil, errors.New("hwloc_topology_set_xml failure")
+		}
+	}
+
+	status = C.hwloc_topology_set_flags(topology, C.HWLOC_TOPOLOGY_FLAG_IO_DEVICES)
+	if status != 0 {
+		return nil, errors.New("hwloc_topology_set_flags failure")
+	}
+
+	status = C.hwloc_topology_load(topology)
+	if status != 0 {
+		return nil, errors.New("hwloc_topology_load failure")
+	}
+	return topology, nil
+}
+
+// cleanUp closes out the hwloc resources
+func cleanUp(topology C.hwloc_topology_t) {
+	if (topology != nil) {
+		C.hwloc_topology_destroy(topology)
+	}
+}
+
+// GetAffinityForNetworkDevices searches the system topology for each network
+// device specified by the input string and returns corresponding cpuset and
+// nodeset bitmaps as strings.  When converted back to hwloc_bitmap_t via
+// hwloc_bitmap_sscanf() these bitmaps are used by the hwloc API to bind a
+// thread to processing units that are closest to the given network device.
+//
+// The input netNames string has the form "device0,device1,...deviceN"
+// where 'device' specifies a network device such as "eth0", "eth1", or "ib0"
+// The return string has the form:
+// "dev0:cpuset:nodeset;dev1:cpuset:nodeset; ... devN:cpuset:nodeset"
```

Comment:
Where did the requirement for the input and output formats as delimited strings come from? It's a little odd to force every consumer to generate a formatted input string and parse out complex information from the output string. What about something like this instead:

```go
type DeviceAffinity struct {
        DeviceName string
        CPUSet     string
        NodeSet    string
}

func (da *DeviceAffinity) String() string {
        return fmt.Sprintf("%s:%s:%s", da.DeviceName, da.CPUSet, da.NodeSet)
}

type DeviceAffinityMap map[string]*DeviceAffinity

func (dam DeviceAffinityMap) String() string {
        deviceStrings := make([]string, 0, len(dam))
        for _, device := range dam {
                deviceStrings = append(deviceStrings, device.String())
        }
        return strings.Join(deviceStrings, ";")
}

func GetAffinityForNetworkDevices(deviceNames ...string) (DeviceAffinityMap, error) {
...
```
This way, consumers can boil it down to a string if they want, or else easily extract specific details from the result. Making deviceNames a variadic argument makes the API friendlier, and you can just use `netnames := strings.Join(deviceNames, ",")` to keep the same logic for matching against topology results.